pkg/ssh/connection_golang.go

```diff
@@ -272,23 +273,54 @@ func ValidateAndConfigure(uri *url.URL, iden string) (*ssh.ClientConfig, error)
 	if err != nil {
 		return nil, err
 	}
-	keyFilePath := filepath.Join(homedir.Get(), ".ssh", "known_hosts")
-	known, err := knownhosts.New(keyFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("creating host key callback function for %s: %w", keyFilePath, err)
-	}
 
+	var keyErr *knownhosts.KeyError
 	cfg := &ssh.ClientConfig{
-		User:            uri.User.Username(),
-		Auth:            authMethods,
-		HostKeyCallback: known,
-		Timeout:         tick,
+		User: uri.User.Username(),
+		Auth: authMethods,
+		HostKeyCallback: ssh.HostKeyCallback(func(host string, remote net.Addr, pubKey ssh.PublicKey) error {
+			keyFilePath := filepath.Join(homedir.Get(), ".ssh", "known_hosts")
+			known, err := knownhosts.New(keyFilePath)
+			if err != nil {
+				if os.IsNotExist(err) && insecureMachine {
+					logrus.Warn("please create a known_hosts file and recreate the machine to add the public key entry")
+					return nil
+				}
+				return err
+			}
+			hErr := known(host, remote, pubKey)
+			if errors.As(hErr, &keyErr) && len(keyErr.Want) > 0 {
+				logrus.Warnf("ssh host key mismatch for host %s, got key %s of type %s", host, ssh.FingerprintSHA256(pubKey), pubKey.Type())
+				return keyErr
+			} else if errors.As(hErr, &keyErr) && len(keyErr.Want) == 0 {
```

Comment:
Could you write down, in words, the intended policy?

This might be correct but it seems surprising. In particular I didn’t expect that a missing `known_hosts`, and an existing `known_hosts` without an entry for the host in question, don’t cause the same behavior. Why does it warn only on a missing file? Why does it add an entry only on an existing file with a missing entry, and not with an empty file?

Is there, maybe, some expectation that in machine mode, we would _never_ get in the `len(keyErr.Want) == 0` situation? If so, this should loudly fail if we do get in that situation.

Reply:
well, since we are adding the machine keys here too, this will fail for machine as well. I am adding comments to explain each condition

Reply:
AFAICS if `known_hosts` does not exist, we either warn and succeed, or don’t warn and fail, but we never get to the `known(` call and never get the opportunity to add a new known_hosts entry.

---

To be a bit more explicit, the “the intended policy” was asking about what is the ~higher-level wanted behavior, maybe I’d phase it as asking what are the guarantees that users can rely on, and what are the compatibility compromises we have decided to use that weaken security or usability.

Explaining the code line by line might not be the same thing — i.e. please focus on _why_ the code is using these conditions to decide, not just re-explain the details of the effects of the Go code in words. At least assuming I’ve been reading the code correctly (which might not be the case), I don’t think we are quite in disagreement in what the code _does_ on any code path — but I find it difficult to figure out what is the overall effect.

Reply:
I have added some intense documentation to this section and the code here is good to go, resolving this.

Reply:
The added line-by-line documentation answers not a single one of the specific questions from this thread:
- I didn’t expect that a missing known_hosts, and an existing known_hosts without an entry for the host in question, don’t cause the same behavior. Why does it warn only on a missing file? Why does it add an entry only on an existing file with a missing entry, and not with an empty file?
- what are the guarantees that users can rely on
- what are the compatibility compromises we have decided to use that weaken security or usability.

Reply:
I will resolve this so we can talk about it. @mtrmac would you be opposed to getting this to a mergeable state but me filing an issue to keep track of the remaining design discussions we want to have?

Reply:
You’re right, let’s collect the outstanding ideas somewhere. https://github.com/containers/common/issues/1152 , feel free to edit / add entries.